### PR TITLE
Parse and insert blocks

### PR DESCRIPTION
**Pull request by OttoAllmendinger**
* Implement `parseBlock()`
* Implement `Storage.InsertBlock()`
* Update transactions and `last_removed` on reorg

### CODE REVIEW

File: src/types/block.go [hunk 1]
```diff
@@ -13,6 +20,72 @@ type Block struct {
 	// TODO: Size ?
 }
 
+// https://bitcoin.org/en/developer-reference#coinbase
+func parseHeight(txin wire.TxIn) (int, error) {
```
Comment (0xB10C) [inline src/types/block.go]: Doesn't need to return an error

File: src/daemon/daemon.go [hunk 2]
```diff
@@ -38,9 +38,9 @@ func (b *BademeisterDaemon) processTransaction(tx *types.Transaction) error {
 }
 
 func (b *BademeisterDaemon) processBlock(block *types.Block) error {
-	log.Printf("Received block, updating transactions")
-	// TODO update storage
-	return nil
+	log.Printf("Received block, updating database")
```
Comment (0xB10C) [inline src/daemon/daemon.go]: nit: could log block hash here

File: src/storage/storage.go [hunk 3]
```diff
@@ -12,6 +13,13 @@ import (
 
 const currentVersion = 4
```
Comment (0xB10C) [inline src/storage/storage.go]: Should we up the database version?

Comment (OttoAllmendinger) [reply src/storage/storage.go]: Yes, good catch. Otherwise the new tables will be missing.

File: src/daemon/daemon.go (hunk 2 above)
Comment (OttoAllmendinger) [reply src/daemon/daemon.go]: Good idea. I have some other logging changes planned and would like to change it there

https://github.com/0xB10C/bademeister-go/issues/35

File: src/types/block.go (hunk 1 above)
Comment (OttoAllmendinger) [reply src/types/block.go]: will fix

File: src/storage/storage_transaction.go [hunk 4]
```diff
@@ -182,6 +182,29 @@ func (s *Storage) transactionDBIDs(txids []types.Hash32) (*[]int64, error) {
...
+			"transaction"
+		WHERE
+			id IN (
+				SELECT
+					transaction_id
+				FROM
+					"transaction_block"
+				WHERE
+					block_id = ?
+			)`, blockId)
```
Comment (0xB10C) [inline src/storage/storage_transaction.go]: Could benchmark if a JOIN is faster. But can be done once we run into performance issues and does not need to be done now.

**Comment by 0xB10C** [comment]
Still want to take another look at `storage` and the tests. Looks good otherwise.

File: src/storage/storage.go [hunk 5]
```diff
@@ -131,6 +139,34 @@ func (s *Storage) initialize(version int) error {
 		return errors.Errorf("could not create the table `transaction`: %s", err)
 	}
 
+	const createBlockTable string = `
+		CREATE TABLE "block" (
+			 id         INTEGER PRIMARY KEY UNIQUE NOT NULL, 
+			 hash       BLOB (32) UNIQUE NOT NULL, 
+			 parent     BLOB (32),
```
[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply src/storage/storage.go]: there would also be additional complexity when tracking the first seen block which has a parent that is not in the DB

**Pull request merged**